src/server/service/user.ts

```diff
@@ -30,7 +29,28 @@ export default class UserService implements IUserService {
     return hash;
   }
 
-  async create(dto: CreateUserDTO): Promise<User> {
+  async create(dto: CreateUserDTO, email: string): Promise<User> {
+    const existentUser = await this.findByEmail(email);
+    if (!dto.role) {
+      throw new Error('"role" is a required field!');
```

Comment:
Let's use dedicated errors for this, we cant rely on generic `Error` instance to check which error occurred here.

For instance, if we wanted to check on this error from the request handler side, we would do something like:

```ts
try {
  // handler logic
} catch (error) {
  // here we have to check for two conditions to know if `error` is an instance of the error from L35
  // 1. We need to check if `error.message` is defined
  // 2. We need to check if `error.message` is strictly equal to a specific string
  if (typeof error.message !== 'undefined' && error.message === '"role" is a required field!') {
    return httpResponse.badRequestMessage('The "role" property is required');
  }
  // -- snip --
}
```

Reply:
I suggest implementing errors for each possible case, for this case you would have something like:

```ts
class InvalidUserRole extends Error {
  constructor() {
    super('The "role" value should be one of: "admin", "manager" or "user"');

    this.name = 'InvalidUserRole';
  }
}
```

With the `InvalidUserRole` class implemented, you can throw an instance of this class instead of the straight `Error` class instance.

```suggestion
      throw new InvalidUserRole();
```

Reply:
Going back to our handler example, the final check would look like the following:

```ts
try {
  // handler logic
} catch (error) {
  if (error instanceof InvalidUserRole) {
    return httpResponse.badRequestMessage(error.message);
  }
  // -- snip --
}
```